Arm/Memory/SeparateAutomation.lean

```diff
@@ -482,10 +585,74 @@ def MemSeparateProof.omega_def (h : MemSeparateProof e) : Expr :=
 def MemSeparateProof.addOmegaFacts (h : MemSeparateProof e) (args : Array Expr) :
     SimpMemM (Array Expr) := do
   SimpMemM.withMainContext do
-    let fvar ← introDef "hmemSeparate_omega" h.omega_def
+    -- simp only [bitvec_rules] (failIfUnchanged := false)
+    let fvar ← simpAndIntroDef "hmemSeparate_omega" h.omega_def
     trace[simp_mem.info]  "{h}: added omega fact ({h.omega_def})"
     return args.push (Expr.fvar fvar)
 
+
+
+/-- info: Ne.{u} {α : Sort u} (a b : α) : Prop -/
+#guard_msgs in #check Ne
+
+/-- info: List.get?.{u} {α : Type u} (as : List α) (i : Nat) : Option α -/
+#guard_msgs in #check List.get?
+
+/-- Make the expression `mems.get? i = some a`. -/
+def mkListGetEqSomeTy (mems : MemPairwiseSeparateProp) (i : Nat) (a : MemSpanExpr) : SimpMemM Expr := do
+  let lhs ← mkAppOptM ``List.get? #[.none, mems.getMemSpanListExpr, mkNatLit i]
+  let rhs ← mkSome MemSpanExpr.toTypeExpr a.toExpr
+  mkEq lhs rhs
+
+/--
+info: Memory.Region.separate'_of_pairwiseSeprate_of_mem_of_mem {mems : List Memory.Region}
+  (h : Memory.Region.pairwiseSeparate mems) (i j : Nat) (hij : i ≠ j) (a b : Memory.Region) (ha : mems.get? i = some a)
+  (hb : mems.get? j = some b) : mem_separate' a.fst a.snd b.fst b.snd
+-/
+#guard_msgs in #check Memory.Region.separate'_of_pairwiseSeprate_of_mem_of_mem
+
+/-- make `Memory.Region.separate'_of_pairwiseSeprate_of_mem_of_mem i j (by decide) a b rfl rfl`. -/
+def MemPairwiseSeparateProof.mem_separate'_of_pairwiseSeparate_of_mem_of_mem
+    (self : MemPairwiseSeparateProof mems) (i j : Nat) (a b : MemSpanExpr)  :
+    SimpMemM <| MemSeparateProof ⟨a, b⟩ := do
+  let iexpr := mkNatLit i
+  let jexpr := mkNatLit j
+    -- i ≠ j
+  let hijTy := mkAppN (mkConst ``Ne [1]) #[(mkConst ``Nat), mkNatLit i, mkNatLit j]
+  -- mems.get? i = some a
+  let hijVal ← mkDecideProof hijTy
+  let haVal ← mkEqRefl <| ← mkSome MemSpanExpr.toTypeExpr a.toExpr
+  let hbVal ← mkEqRefl <| ← mkSome MemSpanExpr.toTypeExpr b.toExpr
+  let h := mkAppN (Expr.const ``Memory.Region.separate'_of_pairwiseSeprate_of_mem_of_mem [])
+    #[mems.getMemSpanListExpr,
+      self.h,
+      iexpr,
+      jexpr,
+      hijVal,
+      a.toExpr,
+      b.toExpr,
+      haVal,
+      hbVal]
+  return ⟨h⟩
+/--
+Currently, if the list is syntacticaly of the form [x1, ..., xn],
+ we create hypotheses of the form `mem_separate' xi xj` for all i, j..
+This can be generalized to pairwise separation given hypotheses x ∈ xs, x' ∈ xs.
+-/
+def MemPairwiseSeparateProof.addOmegaFacts (h : MemPairwiseSeparateProof e) (args : Array Expr) :
```

Comment:
This looks like a lot of omega facts, and maybe that's fine performance-wise, but I'm still nervous.

Can we think about getting these facts smartly/lazily from the context instead? Lot more work, plus we need to make sure we need to be that smart. Thoughts?

Reply:
@shigoel I thought about this a bit, and unfortunately, I convinced myself that there's no complete heuristic for this, so we are forced to know all $$O(n^2)$$ facts in the worst case. Intuitively, we way need to establish something based on knowing the mutual exclusion of *all* memory regions. 

I would like to keep a watch on this, though, and see how it scales as we get larger examples.

Reply:
Agree with the intuition, and this works for now.

I was suggesting a lazy pick-a-point heuristic for the future (if need be, and we don't know that until we try bigger examples, I agree): keep picking a separateness fact from the list of memory regions until the goal can't be simplified any further, in which case we throw all the facts in the context. This could work in the common case; very rarely do we need to establish something from the mutual exclusion of _all_ the regions.

But this could be an awful heuristic because of the back-and-forth between picking a fact and attempted simplification; we could build a way to turn this heuristic off. I don't know. Anyway, just a thought for now.